railties/lib/rails/commands/test/test_command.rb

```diff
@@ -57,6 +60,12 @@ def generators(*)
         args.prepend("test/lib/generators")
         perform
       end
+
+      private
+        def run_prepare_task
+          success = kernel_system("rake", "test:prepare")
```

Comment:
@etiennebarrie given your work on https://github.com/rails/rails/pull/44106 and https://github.com/rails/rails/pull/41132 I'm hoping you can help with this one 🙏  (At the very least it explains why the `test:prepare` task exists.)

Some tests are failing in CI, it looks like they are all cases where we're trying to run `rails test` on a plugin or engine which doesn't have any associated Rake tasks. So the command isn't found, and this raises.

I tried this implementation as an alternative:

```diff
diff --git a/railties/lib/rails/commands/test/test_command.rb b/railties/lib/rails/commands/test/test_command.rb
index 81be58b88f..1c00327d95 100644
--- a/railties/lib/rails/commands/test/test_command.rb
+++ b/railties/lib/rails/commands/test/test_command.rb
@@ -1,6 +1,7 @@
 # frozen_string_literal: true

 require "rails/command"
+require "rails/commands/rake/rake_command"
 require "rails/test_unit/runner"
 require "rails/test_unit/reporter"

@@ -50,7 +51,6 @@ def all(*)
         perform
       end

-      alias kernel_system system
       def system(*)
         args.prepend("test/system")
         perform
@@ -63,8 +63,7 @@ def generators(*)

       private
         def run_prepare_task
-          success = kernel_system("rake", "test:prepare")
-          success || exit(false)
+          Rails::Command::RakeCommand.perform("test:prepare")
         end
     end
   end
```

But I got the error you mention in #44106, where it was changing the environment from test to development when tests run.

Anyway, I'm just wondering if you (or anyone else reading) have any ideas how we could run the Rake task here *only* if it's defined.

Reply:
Rake has `Rake::Task.task_defined?` which you could use. I think it'd be relatively straightforward to add a `Rails::Command::RakeCommand.perform_if_exists(task)` that uses that.

Regarding the environment, it gets set in `Rails::TestUnit::Runner.parse_options`, so you need to call `run_prepare_task` after that to use the test environment (or whichever environment is asked with `--environment`).

Reply:
That tip about `parse_options` is what I was missing. Thank you!!